src/pip/_internal/utils/filesystem.py

```diff
@@ -28,3 +45,33 @@ def check_path_owner(path):
         else:
             previous, path = path, os.path.dirname(path)
     return False  # assume we don't own the path
+
+
+@contextmanager
+def adjacent_tmp_file(path):
+    # type: (str) -> Iterator[NamedTemporaryFileResult]
+    """Given a path to a file, open a temp file next to it securely and ensure
+    it is written to disk after the context reaches its end.
+    """
+    with NamedTemporaryFile(
+        delete=False,
+        dir=os.path.dirname(path),
+        prefix=os.path.basename(path),
+        suffix='.tmp',
+    ) as f:
+        result = cast('NamedTemporaryFileResult', f)
+        try:
+            yield result
+        finally:
+            result.file.flush()
+            os.fsync(result.file.fileno())
+
+
+@retry(stop_max_delay=1000, wait_fixed=250)
+def replace(src, dest):
+    # type: (str, str) -> None
+    try:
+        os.rename(src, dest)
+    except OSError:
+        os.remove(dest)
+        os.rename(src, dest)
```

Comment:
There are quite several implementations floating around for this, and most of them handle some edge case better than another. I wonder whether we should reuse one of them, instead of rolling our own (and having to deal with edge case reports for a while).

Some examples:

* [python-atomicwrites](https://python-atomicwrites.readthedocs.io/en/latest/) has a better `replace` implementation, especially on Windows.
* [Lektor contains a similar implementation.](https://github.com/lektor/lektor/blob/bd6dca51a3a77ed8acacf6ca9303e2913e50f78c/lektor/utils.py#L455) Unfortunately it relies on Werkzeug for robust file replacement, which makes pip intergration quite infeasible.
* [Armin has another implementation](https://github.com/mitsuhiko/python-atomicfile), this one self-contained. Not sure how well it works.
* Trac also has an `AtomicFile` utility.
* [Pipenv uses a stripped-down version of Lektor’s.](https://github.com/sarugaku/vistir/blob/1dcb51f5c39b17b1693992a21e59b78f1719c2bc/src/vistir/contextmanagers.py#L167)
* Python 3.3+ has a built-in `os.replace()` that would be more robust than the `try-rename-remove` dance.

Reply:
I'd be inclined to use `os.replace()` on Python 3, and fall back to what's above (with the understanding that it's limited and we don't expect to worry too much about edge cases) for Python 2.

Reply:
I second @pfmoore.

Let's use `os.replace` on Python 3, and do "enough" on Python 2.

Reply:
Updated [here](https://github.com/pypa/pip/pull/6879/files#diff-6d5635c4f949690f8e6196ae6b155582R102-R115).